### PR TITLE
https://applicaster.atlassian.net/browse/IM-220

### PR DESCRIPTION
**Pull request by RoeiCarmelMsApps**
favourites cross platform support

### CODE REVIEW

File: PluginClasses/ServiceApi.swift [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    let apiName = "\(baseApiUrl)\(favouritePath)\(uid)"
+    let url = URL(string: apiName)!
+    let request = NSMutableURLRequest(url: url)
+    request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
+    request.httpMethod = "GET"
+    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+    makeRequest(request: request) { (responseJson, response, error) in
+        if(response?.statusCode == 200){
+            completion(true)
+        }else if (response?.statusCode == 404){
```
Comment (snooky23) [inline PluginClasses/ServiceApi.swift]: what about not modified?

https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/304

File: PluginClasses/ServiceApi.swift [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+    if(on){
+        let params: [String:String] = ["object_uid" : uid]
+        request.httpBody = try! JSONSerialization.data(withJSONObject: params, options: .prettyPrinted)
+    }
+    makeRequest(request: request) { (responseJson, response, error) in
+        if(response?.statusCode == 200){
+            completion(true)
+        }else if(response?.statusCode == 404){
+            completion(true)
```
Comment (snooky23) [inline PluginClasses/ServiceApi.swift:77]: why are you returning true for 404?

File: PluginClasses/ZPTVACellViewController.swift [hunk 3]
```diff
@@ -102,8 +107,61 @@ class ZPTVACellViewController : CACellViewController {
         }
     }
     
+    private func isFavoriteSupported() -> UIButton?{
+        guard let layout = self.currentComponentModel()?.style["layout_name"] as? String , let favoriteBtn  = favoritesButton, layout == "Family_5_horizontal_list_13" || layout == "Family_5_grid_11" || layout == "Family_5_grid_12"  else{
```
Comment (snooky23) [inline PluginClasses/ZPTVACellViewController.swift]: cont string vars

Comment (snooky23) [reply PluginClasses/ZPTVACellViewController.swift]: every let check should be on a different line

File: PluginClasses/ZPTVACellViewController.swift [hunk 4]
```diff
@@ -102,8 +107,61 @@ class ZPTVACellViewController : CACellViewController {
...
+        }
+        if(sender.isSelected){
+            setFavoriteState(uid: atomFeed.identifier, on: false) { (success) in
+                if(success){
+                    DispatchQueue.onMain {
+                        sender.isSelected = false
+                    }
+                }
+            }
+        }else{
```
Comment (snooky23) [inline PluginClasses/ZPTVACellViewController.swift]: add a space between } and else

File: PluginClasses/ServiceApi.swift [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+//
+
+import Foundation
+
+let tokenKey = "AWSCognitoTokenKey"
+let defaults = UserDefaults.standard
+var baseApi :String?
+let favouritePath = "/userlists/favorites/"
+
+func getFavouriteState(uid: String, completion: @escaping (( _ on: Bool?) -> Void)){
```
Comment (snooky23) [inline PluginClasses/ServiceApi.swift:27]: Please add comments before every method.

File: PluginClasses/ServiceApi.swift [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        }
+    }
+}
+
+func setFavoriteState(uid: String, on: Bool, completion: @escaping (( _ success: Bool) -> Void)){
+    guard let token = defaults.string(forKey: tokenKey), let baseApiUrl = baseApi else{
+        return
+    }
+    let apiName = "\(baseApiUrl)\(favouritePath)\(uid)"
+    let url = URL(string: apiName)!
```
Comment (snooky23) [inline PluginClasses/ServiceApi.swift]: guard it!
Please don't use ! notation.

File: PluginClasses/ServiceApi.swift [hunk 7]
```diff
@@ -0,0 +1,93 @@
+//
+//  ServiceApi.swift
+//  TVADataBinder
+//
+//  Created by MSApps on 16/04/2020.
```
Comment (snooky23) [inline PluginClasses/ServiceApi.swift:5]: Please use the following enum when checking for http codes.
https://ibm-swift.github.io/Kitura-net/Enums/HTTPStatusCode.html

**Review by snooky23 — CHANGES_REQUESTED** [review]

File: PluginClasses/ServiceApi.swift (hunk 5 above)
Comment (RoeiCarmelMsApps) [reply PluginClasses/ServiceApi.swift:27]: done

File: PluginClasses/ZPTVACellViewController.swift (hunk 4 above)
Comment (RoeiCarmelMsApps) [reply PluginClasses/ZPTVACellViewController.swift]: done

File: PluginClasses/ZPTVACellViewController.swift (hunk 3 above)
Comment (RoeiCarmelMsApps) [reply PluginClasses/ZPTVACellViewController.swift]: done

File: PluginClasses/ServiceApi.swift (hunk 2 above)
Comment (RoeiCarmelMsApps) [reply PluginClasses/ServiceApi.swift:77]: Explained on slack

File: PluginClasses/ServiceApi.swift (hunk 1 above)
Comment (RoeiCarmelMsApps) [reply PluginClasses/ServiceApi.swift]: Explained on slack

File: PluginClasses/ServiceApi.swift [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    guard let apiUrl = url else{
+        return
+    }
+    let request = NSMutableURLRequest(url: apiUrl)
+    request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
+    request.httpMethod = on ? "PUT" : "DELETE"
+    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+    if(on){
+        let params: [String:String] = ["object_uid" : uid]
+        request.httpBody = try! JSONSerialization.data(withJSONObject: params, options: .prettyPrinted)
```
Comment (snooky23) [inline PluginClasses/ServiceApi.swift]: please use guard let

File: PluginClasses/ServiceApi.swift [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+            on ? completion(false) : completion(true)
+        } else{
+            completion(false)
+        }
+    }
+}
+
+private func makeRequest(request: NSMutableURLRequest, completion: @escaping ((_ result: Any?, _ httpResponse: HTTPURLResponse?, _ error: Error?) -> Void)) {
+    let task = URLSession.shared.dataTask(with: request as URLRequest) { (data, response, error) in
+        guard let data = data , let json = (try? JSONSerialization.jsonObject(with: data, options: [])) else {
```
Comment (snooky23) [inline PluginClasses/ServiceApi.swift]: every check should be in a different line

File: PluginClasses/ZPTVACellViewController.swift [hunk 10]
```diff
@@ -102,8 +107,64 @@ class ZPTVACellViewController : CACellViewController {
         }
     }
     
+    private func isFavoriteSupported() -> UIButton?{
+        guard let layout = self.currentComponentModel()?.style["layout_name"] as? String,
+            let favoriteBtn  = favoritesButton,
+            layout == "Family_5_horizontal_list_13" || layout == "Family_5_grid_11" || layout == "Family_5_grid_12"  else{
```
Comment (snooky23) [inline PluginClasses/ZPTVACellViewController.swift]: Please filter to check all values in an array https://stackoverflow.com/questions/30828074/check-if-array-contains-part-of-a-string-in-swift

File: PluginClasses/ZPTVADataBinder.swift [hunk 11]
```diff
@@ -14,10 +14,17 @@ class ZPDataBinderTVA: ZPGeneralBaseProvider, ZPGeneralPluginUIProtocol {
...
+        guard let config = configurationJSON as? [String:Any] else{
+            return
+        }
+        if(styleArry == nil){
+            getStringJsonFromConfig(config: config)
+        }
+        
+        if let baseApiUrl = config["favorites_service_base_url"] as? String{
+            baseApi = baseApiUrl
+        }else{
```
Comment (snooky23) [inline PluginClasses/ZPTVADataBinder.swift]: add a space between else and }

File: PluginClasses/ServiceApi.swift [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+        return
+    }
+    let request = NSMutableURLRequest(url: apiUrl)
+    request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
+    request.httpMethod = "GET"
+    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+    makeRequest(request: request) { (responseJson, response, error) in
+        if(response?.statusCode == HTTPStatusCode.success.rawValue){
+            completion(true)
+        }else if (response?.statusCode == HTTPStatusCode.notFound.rawValue){
```
Comment (snooky23) [inline PluginClasses/ServiceApi.swift]: add a space before the else check

File: PluginClasses/ServiceApi.swift [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+            completion(false)
+        }else{
```
Comment (snooky23) [inline PluginClasses/ServiceApi.swift]: add a space before the else check

File: PluginClasses/ServiceApi.swift [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+        return
+    }
+    let apiName = "\(baseApiUrl)\(favouritePath)\(uid)"
+    let url = URL(string: apiName)
+    guard let apiUrl = url else{
+        return
+    }
+    let request = NSMutableURLRequest(url: apiUrl)
+    request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
+    request.httpMethod = on ? "PUT" : "DELETE"
```
Comment (snooky23) [inline PluginClasses/ServiceApi.swift]: change the on var to httpMethod

File: PluginClasses/ServiceApi.swift [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
+    if(on){
+        let params: [String:String] = ["object_uid" : uid]
+        request.httpBody = try! JSONSerialization.data(withJSONObject: params, options: .prettyPrinted)
+    }
+    makeRequest(request: request) { (responseJson, response, error) in
+        if(response?.statusCode == HTTPStatusCode.success.rawValue){
+            completion(true)
+        }else if(response?.statusCode == HTTPStatusCode.notFound.rawValue){
```
Comment (snooky23) [inline PluginClasses/ServiceApi.swift]: use filter method https://stackoverflow.com/questions/30828074/check-if-array-contains-part-of-a-string-in-swift

**Review by snooky23 — CHANGES_REQUESTED** [review]

File: PluginClasses/ServiceApi.swift (hunk 15 above)
Comment (RoeiCarmelMsApps) [reply PluginClasses/ServiceApi.swift]: done

File: PluginClasses/ServiceApi.swift (hunk 14 above)
Comment (RoeiCarmelMsApps) [reply PluginClasses/ServiceApi.swift]: done

File: PluginClasses/ServiceApi.swift (hunk 13 above)
Comment (RoeiCarmelMsApps) [reply PluginClasses/ServiceApi.swift]: done

File: PluginClasses/ServiceApi.swift (hunk 12 above)
Comment (RoeiCarmelMsApps) [reply PluginClasses/ServiceApi.swift]: done

File: PluginClasses/ZPTVADataBinder.swift (hunk 11 above)
Comment (RoeiCarmelMsApps) [reply PluginClasses/ZPTVADataBinder.swift]: done

File: PluginClasses/ZPTVACellViewController.swift (hunk 10 above)
Comment (RoeiCarmelMsApps) [reply PluginClasses/ZPTVACellViewController.swift]: done

File: PluginClasses/ServiceApi.swift (hunk 8 above)
Comment (RoeiCarmelMsApps) [reply PluginClasses/ServiceApi.swift]: done

File: PluginClasses/ServiceApi.swift (hunk 9 above)
Comment (RoeiCarmelMsApps) [reply PluginClasses/ServiceApi.swift]: done

**Review by snooky23 — APPROVED** [review]

**Pull request merged**